### PR TITLE
align ClassStaticBlockDeclaration with IIFE in CFA

### LINKED ISSUE #44949: Statements in class static blocks do not narrow types
# Bug Report

### 🔎 Search Terms

control flow analysis, class static blocks

### 🕗 Version & Regression Information

- v4.4.0-beta
- v4.4.0-dev.20210708
- This is the behavior in every version I tried, and I reviewed the [proposal](https://github.com/tc39/proposal-class-static-block/blob/main/README.md)
- I was unable to test this on prior versions because class static blocks were added in 4.4.0

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.4.0-dev.20210708#code/DYUwLgBAZg9jBcEB2BXAtgIxAJwNwFgAoIgY2AEMBnSiAYQgG8iIJKxywBLEx5l6OBAC8EAMwFCLAL5EZxeSRhJKMUADpgMAOYAKWDACUEoA)

### 💻 Code

```ts
let foo: number;

class C {
  static {
    foo = 3;
  }
}

console.log(foo);
```

### 🙁 Actual behavior

An error `Variable 'foo' is used before being assigned.(2454)` occurs at line 10, `console.log(foo)`.

### 🙂 Expected behavior

Code compiles without errors.

### LINKED ISSUE #45932: Class static block does not allow forward reference to static method
# Bug Report

### 🔎 Search Terms

class static blocks, forward reference, incorrect compiler error

### 🕗 Version & Regression Information

4.4.2 + nightly (4.5.0-dev20210910)

- I was unable to test this on prior versions because the feature is new in 4.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.4.2#code/MYGwhgzhAECC0G8CwAoa7oQC5iwS2EVQxLgDoATAewGUqBbAUywAs8A7AcwAoBKAbmgB6IdACiAJwlUJALmgAiAArSADowlYAntADk1Ok1YdOu6HhgBXCIwrQARowBmMxuawwOefGBB4AXrh4VOxk3ABMAOzhAJy8CsQYAL6oiejYQYQGDMxsXHxEaKTQwCEQVCCMZCBUPAqcVFjALGAAhAoCadApKElAA)

### 💻 Code

```ts
class A {
    static {
        A.doSomething();
    }

    static doSomething() {
       console.log("gotcha!");
    }
}
```

### 🙁 Actual behavior

Compiler error in line 3: "Property 'doSomething' is used before its initialization.(2729)"

The code works as-is in native  JavaScript, tested in Firefox Developer Edition.

The rewrites for class static blocks in Babel as well as in TypeScript itself also work. Both execute the static block's code at a time where static methods of the class are already available.

### 🙂 Expected behavior

No compiler error, because everything's fine.

A forward reference to a static _field_, however, should be an error (I guess).

### Workaround

Fall back to the way class static blocks could be emulated before they were supported:
Define an unused #-private static constant whose initializer is an immediate-evaluating function.

```ts
class A {
    static #_ = (() => {
        A.doSomething();
    })();

    static doSomething() {
       console.log("gotcha!");
    }
}
```

[Playground link for workaround](https://www.typescriptlang.org/play?ts=4.4.2#code/MYGwhgzhAECC0G8CwAoa7oQC5iwS2GgGIB9aAXmgAoqBKCgPkVQ1bgDoATAewGVuAtgFMsACzwA7AOZ0A3CwwBfWnNQL02XAWg9+wsZJn1kaNtGDcJEbiCHsQ3GQCIp3LMFFgAhE9rzT6IqoikA)

Looks ugly, but works in the desired execution order and produces no compiler error.

### Possibly related issues

- Execution order of static field initialization and static initialization blocks is not kept under target: esnext and useDefineForClassFields: false #45574
- Statements in class static blocks do not narrow types #44949

### PR DESCRIPTION
**Pull request by Zzzen**
Fixes #44949
Fixes #45932

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -657,19 +657,23 @@ namespace ts {
...
                 const saveActiveLabelList = activeLabelList;
                 const saveHasExplicitReturn = hasExplicitReturn;
-                const isIIFE = containerFlags & ContainerFlags.IsFunctionExpression && !hasSyntacticModifier(node, ModifierFlags.Async) &&
-                    !(node as FunctionLikeDeclaration).asteriskToken && !!getImmediatelyInvokedFunctionExpression(node);
+                const isIIFELike =
+                    (containerFlags & ContainerFlags.IsFunctionExpression &&
+                        !hasSyntacticModifier(node, ModifierFlags.Async) &&
+                        !(node as FunctionLikeDeclaration).asteriskToken &&
+                        !!getImmediatelyInvokedFunctionExpression(node)) ||
+                    node.kind === SyntaxKind.ClassStaticBlockDeclaration;
```
Comment (Zzzen) [inline src/compiler/binder.ts:665]: I am not sure how to format these lines, so here is how prettier handles it.

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -657,19 +657,23 @@ namespace ts {
                 const saveExceptionTarget = currentExceptionTarget;
                 const saveActiveLabelList = activeLabelList;
                 const saveHasExplicitReturn = hasExplicitReturn;
-                const isIIFE = containerFlags & ContainerFlags.IsFunctionExpression && !hasSyntacticModifier(node, ModifierFlags.Async) &&
-                    !(node as FunctionLikeDeclaration).asteriskToken && !!getImmediatelyInvokedFunctionExpression(node);
+                const isIIFELike =
```
Comment (sandersn) [inline src/compiler/binder.ts:660]:
```suggestion
                const isImmediatelyInvoked =
```

(or `isImmediate` for even more old-school flavor)

File: src/compiler/binder.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/binder.ts:665]: I would have done it slightly differently, but prettier's format is fine.

**Review by sandersn — APPROVED** [review]
Looks good but one naming suggestion and it would be nice to have a second opinion on the semantics from @rbuckton

**Comment by fwienber** [comment]
Hi! I just left a [new test case at the issue](https://github.com/microsoft/TypeScript/issues/44949#issuecomment-918208746) which you might want to check! Would be awesome if that case was covered by your fix, too!

**Comment by fwienber** [comment]
Anything missing here but the last review? I'd love to retest this in the nightly build as soon as it is merged!

**Comment by Zzzen** [comment]
@rbuckton please take a look.

**Review by rbuckton — APPROVED** [review]

**Comment by rbuckton** [comment]
I'll have the bot run some extensive tests to make sure there's no other side effects from this change.

**Comment by rbuckton** [comment]
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

**Comment by rbuckton** [comment]
None of the reported test failures look to be related to this change. I think this is good to merge.

**Comment by fwienber** [comment]
Sorry to be unfamiliar with your process, but could you tell me how I can see which milestone / release this fix will be in?
It does not seem to be included in the [4.7.0 Milestone](https://github.com/microsoft/TypeScript/milestone/156) (yet?).

**Comment by fwienber** [comment]
And also not part of the latest nightly build (if I didn't do something wrong when trying to test with the nightly)...

**Comment by Zzzen** [comment]
Did you tried [4.7.0-dev.20220329](https://www.npmjs.com/package/typescript/v/4.7.0-dev.20220329)?

**Comment by fwienber** [comment]
My reproducer is now also fixed in TypeScript playground when choosing "nightly" ☺

**Pull request merged**